### PR TITLE
Vendor helm chart versions based on chartfile.yaml

### PR DESCRIPTION
**Pull request by craigfurman**
When calling `tk tool charts vendor`, and a directory already exists at
the expected chart path, recursively delete that directory unless its
Chart.yaml matches the version of the chart specified in chartfile.yaml.

This allows vendored charts whose versions have changed to be downloaded
by tanka users without first removing the charts directory.

Return errors when Chart.yaml doesn't exist, or is malformed.

---

Implements https://github.com/grafana/tanka/issues/419

### CODE REVIEW

File: pkg/helm/charts.go
Comment (craigfurman) [inline pkg/helm/charts.go:121]: RemoveAll is a bit scary. I'm reasonably happy that this code shouldn't delete anyone's entire project checkout (or home dir) unless they configure their chartfile directory to ".", or "../..", and that directory happens to contain a Chart.yaml. Perhaps it's still worth including a more stringent check? What do you reckon?

**Comment by craigfurman** [comment]
@sh0rez would you mind reviewing this implementation of https://github.com/grafana/tanka/issues/419, if you have time?

File: pkg/helm/charts.go
Comment (sh0rez) [inline pkg/helm/charts.go:104]:
```suggestion
				return fmt.Errorf("reading chart manifest: %w", err)
```

Let's omit `error` here, as this error might be wrapped further.

Also we generally use `%w` for errors (go 1.13 error wrapping)

File: pkg/helm/charts.go
Comment (sh0rez) [inline pkg/helm/charts.go:106]: Can we define a struct for this one? Makes unmarshalling and using the result safer

File: pkg/helm/charts.go
Comment (sh0rez) [inline pkg/helm/charts.go:112]: Garbage in the charts folder would make us stop here, with on clear path to fix this.

Either we require manual user interaction (`rm -r` the folder, which we should suggest in that case), or we do it ourselves. 

wdyt? (also /cc @sdboyer, you probably have an opinion here :stuck_out_tongue_closed_eyes:)

**Review by sh0rez — CHANGES_REQUESTED** [review]
Nice work! Sorry it took so long on my end to review this. Feel free to ping me more frequently if this happens again.

I left a couple of comments on details, but I'm generally very happy with this :)

File: pkg/helm/charts.go
Comment (craigfurman) [reply pkg/helm/charts.go:112]: there are 3 error branches here that deal with garbage in the charts folder. On the one hand, we could consider the charts folder to be fully managed by tanka, and do what we want with it. On the other hand, it's always safer to not delete something when you don't know what it is 🙂 

I'm happy to defer to the maintainers on this.

File: pkg/helm/charts.go
Comment (craigfurman) [reply pkg/helm/charts.go]: This produces output like this:

```
Syncing Repositories ...
Pulling Charts ...
reading chart manifest: open /Users/craig/workspace/gitlab.com/gitlab-com/gl-infra/k8s-workloads/tanka-deployments/charts/memcached/Chart.yaml: no such file or directory
```

This error doesn't seem to be wrapped further. If we don't want to prepand "error" here, we could add wrapping higher up the call stack? That might duplicate the word "error" from other branches, if their libraries include that word in error messages though.

File: pkg/helm/charts.go
Comment (craigfurman) [reply pkg/helm/charts.go]: this error branch is now removed as a result of implementing the above request to create a struct for chartManifest, but I think this conversation still applies to the other branches

File: pkg/helm/charts.go
Comment (craigfurman) [reply pkg/helm/charts.go]: done 🙂

**Comment by craigfurman** [comment]
Thanks for reviewing @sh0rez! I've implemented some of the suggestions and replied to another. Let me know what you think 👍

File: pkg/helm/charts.go
Comment (sh0rez) [inline pkg/helm/charts.go:120]: Could this "Removing" also include the reason? ie. we remove it because it has the wrong version, and will be re-downloaded shortly

**Review by sh0rez — APPROVED** [review]
LGTM and sorry you were left without a response for so long.

I like the direction this takes a lot, and while we can surely make it smoother in the future (error messages, failure handling), this is a leap in the right direction. Thanks!

File: pkg/helm/charts.go
Comment (sh0rez) [reply pkg/helm/charts.go:120]: Well done myself, I merged it :stuck_out_tongue_closed_eyes:. Guess this is resolved then

**Comment by craigfurman** [comment]
Thanks @sh0rez! Looks like there are some dangling threads after the merge. I could address them in a follow-on, or we could let them lie for now?

**Pull request merged**
